pkg/resource/instance_profile/sdk.go

```diff
@@ -0,0 +1,450 @@
+	_ = &svcsdk.IAM{}
+	_ = &svcapitypes.InstanceProfile{}
+	_ = ackv1alpha1.AWSAccountID("")
+	_ = &ackerr.NotFound
+	_ = &ackcondition.NotManagedMessage
+	_ = &reflect.Value{}
+	_ = fmt.Sprintf("")
+	_ = &ackrequeue.NoRequeue{}
+)
+
+// sdkFind returns SDK-specific information about a supplied resource
+func (rm *resourceManager) sdkFind(
+	ctx context.Context,
+	r *resource,
+) (latest *resource, err error) {
+	rlog := ackrtlog.FromContext(ctx)
+	exit := rlog.Trace("rm.sdkFind")
+	defer func() {
+		exit(err)
+	}()
+	// If any required fields in the input shape are missing, AWS resource is
+	// not created yet. Return NotFound here to indicate to callers that the
+	// resource isn't yet created.
+	if rm.requiredFieldsMissingFromReadOneInput(r) {
+		return nil, ackerr.NotFound
+	}
+
+	input, err := rm.newDescribeRequestPayload(r)
+	if err != nil {
+		return nil, err
+	}
+
+	var resp *svcsdk.GetInstanceProfileOutput
+	resp, err = rm.sdkapi.GetInstanceProfileWithContext(ctx, input)
+	rm.metrics.RecordAPICall("READ_ONE", "GetInstanceProfile", err)
+	if err != nil {
+		if reqErr, ok := ackerr.AWSRequestFailure(err); ok && reqErr.StatusCode() == 404 {
+			return nil, ackerr.NotFound
+		}
+		if awsErr, ok := ackerr.AWSError(err); ok && awsErr.Code() == "NoSuchEntity" {
+			return nil, ackerr.NotFound
+		}
+		return nil, err
+	}
+
+	// Merge in the information we read from the API call above to the copy of
+	// the original Kubernetes object we passed to the function
+	ko := r.ko.DeepCopy()
+
+	if ko.Status.ACKResourceMetadata == nil {
+		ko.Status.ACKResourceMetadata = &ackv1alpha1.ResourceMetadata{}
+	}
+	if resp.InstanceProfile.Arn != nil {
+		arn := ackv1alpha1.AWSResourceName(*resp.InstanceProfile.Arn)
+		ko.Status.ACKResourceMetadata.ARN = &arn
+	}
+	if resp.InstanceProfile.CreateDate != nil {
+		ko.Status.CreateDate = &metav1.Time{*resp.InstanceProfile.CreateDate}
+	} else {
+		ko.Status.CreateDate = nil
+	}
+	if resp.InstanceProfile.InstanceProfileId != nil {
+		ko.Status.InstanceProfileID = resp.InstanceProfile.InstanceProfileId
+	} else {
+		ko.Status.InstanceProfileID = nil
+	}
+	if resp.InstanceProfile.InstanceProfileName != nil {
+		ko.Spec.Name = resp.InstanceProfile.InstanceProfileName
+	} else {
+		ko.Spec.Name = nil
+	}
+	if resp.InstanceProfile.Path != nil {
+		ko.Spec.Path = resp.InstanceProfile.Path
+	} else {
+		ko.Spec.Path = nil
+	}
+	if resp.InstanceProfile.Tags != nil {
+		f5 := []*svcapitypes.Tag{}
+		for _, f5iter := range resp.InstanceProfile.Tags {
+			f5elem := &svcapitypes.Tag{}
+			if f5iter.Key != nil {
+				f5elem.Key = f5iter.Key
+			}
+			if f5iter.Value != nil {
+				f5elem.Value = f5iter.Value
+			}
+			f5 = append(f5, f5elem)
+		}
+		ko.Spec.Tags = f5
+	} else {
+		ko.Spec.Tags = nil
+	}
+
+	rm.setStatusDefaults(ko)
+
+	// Get the existing role associated with the instance profile if
+	// any. This value needs to later be compared with the user
+	// specified value to make sure they are in sync
+	ko.Spec.Role = nil
```

Comment:
```suggestion
```

Reply:
@a-hilaly So I think this piece of line is actually necessary:

Let's say as an example we have an instance profile that has no role attached. Now let's say we update this profile by adding a new role to the profile.  

In this case, we would actually never hit the below conditional statement (i.e. since the existing profile has no role attached) and won't end up assigning a value to `ko.Spec.Role`. At this point in time, the value of `ko.Spec.Role` would just be a deep copy of our desired spec template (i.e. the role we would like to update to). This ends up in a scenario where the ACK controller decides that there's nothing to reconcile as the deep copy already matches the desired spec, and would mean that the role never gets attached.

`ko.Spec.Role = nil` in this case actually provides an accurate picture of an existing profile having no role attached, and it would make sense for `latest` to reflect this state. I think the code might have logically made more sense if it were:
```
	attachedRoles := resp.InstanceProfile.Roles
	if len(attachedRoles) == 0 {
                 ko.Spec.Role = nil
	} else {
                 ko.Spec.Role = attachedRoles[0].RoleName
        }
``` 
^ But this can actually be shortened to what we currently have. I'll add some comments around this but I'm also happy to switch the code to the if/else statement above if it makes more sense.

Reply:
Ohhh makes sense! Took me some time to wrap my head around it. Looks like the `resp -> resource` transformation isn't doing that for us.